pkg/login/social/azuread_oauth.go

```diff
@@ -0,0 +1,116 @@
+package social
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/models"
+
+	"golang.org/x/oauth2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+type SocialAzureAD struct {
+	*SocialBase
+	allowedDomains []string
+	allowSignup    bool
+}
+
+type azureClaims struct {
+	Email      string   `json:"email"`
+	UniqueName string   `json:"unique_name"`
+	Upn        string   `json:"upn"`
+	Roles      []string `json:"roles"`
+	Name       string   `json:"name"`
+	ID         string   `json:"oid"`
+}
+
+func (s *SocialAzureAD) Type() int {
+	return int(models.AZUREAD)
+}
+
+func (s *SocialAzureAD) IsEmailAllowed(email string) bool {
+	return isEmailAllowed(email, s.allowedDomains)
+}
+
+func (s *SocialAzureAD) IsSignupAllowed() bool {
+	return s.allowSignup
+}
+
+func (s *SocialAzureAD) UserInfo(client *http.Client, token *oauth2.Token) (*BasicUserInfo, error) {
+	idToken := token.Extra("id_token")
+	if idToken == nil {
+		return nil, fmt.Errorf("No id_token found")
+	}
+
+	parsedToken, err := jwt.ParseSigned(idToken.(string))
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing id token")
+	}
+
+	var claims azureClaims
+	if err := parsedToken.UnsafeClaimsWithoutVerification(&claims); err != nil {
+		return nil, fmt.Errorf("Error getting claims from id token")
+	}
+
+	email := extractEmail(claims)
+
+	if email == "" {
+		return nil, errors.New("Error getting user info: No email found in access token")
+	}
+
+	role := s.extractRole(claims)
+
+	return &BasicUserInfo{
+		Id:    claims.ID,
+		Name:  claims.Name,
+		Email: email,
+		Login: email,
+		Role:  string(role),
+	}, nil
+}
+
+func extractEmail(claims azureClaims) string {
+
+	if claims.Email == "" {
+		if len(claims.Upn) > 0 {
+			return claims.Upn
+		}
+		if len(claims.UniqueName) > 0 {
```

Comment:
As mentioned above, it's better to replace `unique_name` with `preferred_username` since it's issued in v2.0 tokens (or add `preferred_username` as additional field to check).

Reply:
I will change it to `preferred_username`.

Reply:
Changed